File: python/sql_injection/product_queries.py
import sqlite3
import psycopg2


# VULN 1: String concatenation SQL injection - product search
def search_products(search_term):
    conn = sqlite3.connect("shopist.db")
    cursor = conn.cursor()
    query = "SELECT id, name, price, stock FROM products WHERE name LIKE '%" + search_term + "%' OR description LIKE '%" + search_term + "%'"
    cursor.execute(query)
    return cursor.fetchall()


# VULN 2: f-string SQL injection - price range filter
def get_products_by_price_range(min_price, max_price):
    conn = psycopg2.connect(host="db", dbname="shopist", user="app", password="apppass")
    cursor = conn.cursor()
    query = f"SELECT * FROM products WHERE price BETWEEN {min_price} AND {max_price} ORDER BY price ASC"
    cursor.execute(query)
    return cursor.fetchall()


# VULN 3: %-format SQL injection with ORDER BY - category filter
def get_products_by_category(category, sort_field):
    conn = sqlite3.connect("shopist.db")
    cursor = conn.cursor()
    query = "SELECT * FROM products WHERE category = '%s' ORDER BY %s" % (category, sort_field)
    cursor.execute(query)
    return cursor.fetchall()

File: go/sql_injection/product_queries.go
package sqlinjection

import (
	"database/sql"
	"fmt"
)

// VULN 1: String concatenation SQL injection - product search
func SearchProducts(searchTerm string) ([]Product, error) {
	query := "SELECT id, name, price, stock FROM products WHERE name LIKE '%" +
		searchTerm + "%' OR description LIKE '%" + searchTerm + "%'"
	rows, err := db.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var products []Product
	for rows.Next() {
		var p Product
		rows.Scan(&p.ID, &p.Name, &p.Price, &p.Stock)
		products = append(products, p)
	}
	return products, nil
}

// VULN 2: fmt.Sprintf SQL injection - price range filter
func GetProductsByPriceRange(minPrice, maxPrice string) ([]Product, error) {
	query := fmt.Sprintf(
		"SELECT * FROM products WHERE price BETWEEN %s AND %s ORDER BY price ASC",
		minPrice, maxPrice,
	)
	rows, err := db.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var products []Product
	for rows.Next() {
		var p Product
		rows.Scan(&p.ID, &p.Name, &p.Price)
		products = append(products, p)
	}
	return products, nil
}

// VULN 3: String concatenation with ORDER BY injection - category filter
func GetProductsByCategory(category, sortField string) ([]Product, error) {
	query := "SELECT * FROM products WHERE category = '" + category + "' ORDER BY " + sortField
	rows, err := db.Query(query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var products []Product
	for rows.Next() {
		var p Product
		rows.Scan(&p.ID, &p.Name, &p.Price, &p.Category)
		products = append(products, p)
	}
	return products, nil
}

// Product placeholder to satisfy compiler
type Product struct {
	ID       int
	Name     string
	Price    float64
	Stock    int
	Category string
}

// User placeholder
type User struct {
	ID       int
	Username string
	Name     string
	Email    string
	Role     string
}

// db placeholder
var _ *sql.DB
